File: backend/src/modules/drouting/drouting.controller.ts
import { Controller, Get, Post, Body, Patch, Param, Delete, Query, ParseIntPipe } from '@nestjs/common';
import { ApiTags, ApiOperation } from '@nestjs/swagger';
import { DroutingService } from './drouting.service';
import { PaginationDto } from '../../common/dto/pagination.dto';
import { ApiResponseDto } from '../../common/dto/api-response.dto';

@ApiTags('动态路由')
@Controller('api/drouting')
export class DroutingController {
  constructor(private readonly droutingService: DroutingService) {}

  // ========== Gateway ==========
  @Get('gateways')
  @ApiOperation({ summary: '获取网关列表' })
  async findAllGateways(@Query() dto: PaginationDto): Promise<ApiResponseDto> {
    const result = await this.droutingService.findAllGateways(dto);
    return ApiResponseDto.success(result);
  }

  @Post('gateways')
  @ApiOperation({ summary: '创建网关' })
  async createGateway(@Body() data: any): Promise<ApiResponseDto> {
    const gateway = await this.droutingService.createGateway(data);
    return ApiResponseDto.success(gateway, '创建成功');
  }

  @Patch('gateways/:id')
  @ApiOperation({ summary: '更新网关' })
  async updateGateway(@Param('id', ParseIntPipe) id: number, @Body() data: any): Promise<ApiResponseDto> {
    const gateway = await this.droutingService.updateGateway(id, data);
    return ApiResponseDto.success(gateway, '更新成功');
  }

  @Delete('gateways/:id')
  @ApiOperation({ summary: '删除网关' })
  async removeGateway(@Param('id', ParseIntPipe) id: number): Promise<ApiResponseDto> {
    await this.droutingService.removeGateway(id);
    return ApiResponseDto.success(null, '删除成功');
  }

  // ========== Rule ==========
  @Get('rules')
  @ApiOperation({ summary: '获取规则列表' })
  async findAllRules(@Query() dto: PaginationDto): Promise<ApiResponseDto> {
    const result = await this.droutingService.findAllRules(dto);
    return ApiResponseDto.success(result);
  }

  @Post('rules')
  @ApiOperation({ summary: '创建规则' })
  async createRule(@Body() data: any): Promise<ApiResponseDto> {
    const rule = await this.droutingService.createRule(data);
    return ApiResponseDto.success(rule, '创建成功');
  }

  @Patch('rules/:id')
  @ApiOperation({ summary: '更新规则' })
  async updateRule(@Param('id', ParseIntPipe) id: number, @Body() data: any): Promise<ApiResponseDto> {
    const rule = await this.droutingService.updateRule(id, data);
    return ApiResponseDto.success(rule, '更新成功');
  }

  @Delete('rules/:id')
  @ApiOperation({ summary: '删除规则' })
  async removeRule(@Param('id', ParseIntPipe) id: number): Promise<ApiResponseDto> {
    await this.droutingService.removeRule(id);
    return ApiResponseDto.success(null, '删除成功');
  }

  // ========== Group ==========
  @Get('groups')
  @ApiOperation({ summary: '获取分组列表' })
  async findAllGroups(@Query() dto: PaginationDto): Promise<ApiResponseDto> {
    const result = await this.droutingService.findAllGroups(dto);
    return ApiResponseDto.success(result);
  }

  @Post('groups')
  @ApiOperation({ summary: '创建分组' })
  async createGroup(@Body() data: any): Promise<ApiResponseDto> {
    const group = await this.droutingService.createGroup(data);
    return ApiResponseDto.success(group, '创建成功');
  }

  @Patch('groups/:id')
  @ApiOperation({ summary: '更新分组' })
  async updateGroup(@Param('id', ParseIntPipe) id: number, @Body() data: any): Promise<ApiResponseDto> {
    const group = await this.droutingService.updateGroup(id, data);
    return ApiResponseDto.success(group, '更新成功');
  }

  @Delete('groups/:id')
  @ApiOperation({ summary: '删除分组' })
  async removeGroup(@Param('id', ParseIntPipe) id: number): Promise<ApiResponseDto> {
    await this.droutingService.removeGroup(id);
    return ApiResponseDto.success(null, '删除成功');
  }

  // ========== Carrier ==========
  @Get('carriers')
  @ApiOperation({ summary: '获取运营商列表' })
  async findAllCarriers(@Query() dto: PaginationDto): Promise<ApiResponseDto> {
    const result = await this.droutingService.findAllCarriers(dto);
    return ApiResponseDto.success(result);
  }

  @Post('carriers')
  @ApiOperation({ summary: '创建运营商' })
  async createCarrier(@Body() data: any): Promise<ApiResponseDto> {
    const carrier = await this.droutingService.createCarrier(data);
    return ApiResponseDto.success(carrier, '创建成功');
  }

  @Patch('carriers/:id')
  @ApiOperation({ summary: '更新运营商' })
  async updateCarrier(@Param('id', ParseIntPipe) id: number, @Body() data: any): Promise<ApiResponseDto> {
    const carrier = await this.droutingService.updateCarrier(id, data);
    return ApiResponseDto.success(carrier, '更新成功');
  }

  @Delete('carriers/:id')
  @ApiOperation({ summary: '删除运营商' })
  async removeCarrier(@Param('id', ParseIntPipe) id: number): Promise<ApiResponseDto> {
    await this.droutingService.removeCarrier(id);
    return ApiResponseDto.success(null, '删除成功');
  }

  // ========== RPC ==========
  @Post('reload')
  @ApiOperation({ summary: '重载动态路由' })
  async reload(): Promise<ApiResponseDto> {
    await this.droutingService.reload();
    return ApiResponseDto.success(null, '重载成功');
  }

  @Get('stats')
  @ApiOperation({ summary: '获取统计信息' })
  async getStats(): Promise<ApiResponseDto> {
    const stats = await this.droutingService.getStats();
    return ApiResponseDto.success(stats);
  }

  @Get('gw-status')
  @ApiOperation({ summary: '获取网关实时状态' })
  async getGwStatus(): Promise<ApiResponseDto> {
    return ApiResponseDto.success(await this.droutingService.getGwStatus());
  }

  @Get('carrier-status')
  @ApiOperation({ summary: '获取运营商实时状态' })
  async getCarrierStatus(): Promise<ApiResponseDto> {
    return ApiResponseDto.success(await this.droutingService.getCarrierStatus());
  }
}

File: kam-admin-module-routing/src/main/java/io/kamailio/admin/modules/drouting/controller/DroutingController.java
package io.kamailio.admin.modules.drouting.controller;

import io.kamailio.admin.common.dto.ApiResponse;
import io.kamailio.admin.common.dto.PaginatedResult;
import io.kamailio.admin.common.dto.PaginationDto;
import io.kamailio.admin.modules.drouting.entity.DrCarrier;
import io.kamailio.admin.modules.drouting.entity.DrGateway;
import io.kamailio.admin.modules.drouting.entity.DrGroup;
import io.kamailio.admin.modules.drouting.entity.DrRule;
import io.kamailio.admin.modules.drouting.service.DroutingService;
import io.swagger.v3.oas.annotations.Operation;
import io.swagger.v3.oas.annotations.tags.Tag;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;

import java.util.Map;

@Tag(name = "动态路由")
@RestController
@RequestMapping("/api/drouting")
public class DroutingController {
    private final DroutingService service;

    public DroutingController(DroutingService service) { this.service = service; }

    @GetMapping("/gateways") public ResponseEntity<ApiResponse<PaginatedResult<DrGateway>>> findAllGateways(PaginationDto dto) { return ResponseEntity.ok(ApiResponse.success(service.findAllGateways(dto))); }
    @PostMapping("/gateways") public ResponseEntity<ApiResponse<DrGateway>> createGateway(@RequestBody DrGateway data) { return ResponseEntity.ok(ApiResponse.success(service.createGateway(data), "创建成功")); }
    @PatchMapping("/gateways/{id}") public ResponseEntity<ApiResponse<DrGateway>> updateGateway(@PathVariable Integer id, @RequestBody DrGateway data) { return ResponseEntity.ok(ApiResponse.success(service.updateGateway(id, data), "更新成功")); }
    @DeleteMapping("/gateways/{id}") public ResponseEntity<ApiResponse<Void>> removeGateway(@PathVariable Integer id) { service.removeGateway(id); return ResponseEntity.ok(ApiResponse.success(null, "删除成功")); }

    @GetMapping("/rules") public ResponseEntity<ApiResponse<PaginatedResult<DrRule>>> findAllRules(PaginationDto dto) { return ResponseEntity.ok(ApiResponse.success(service.findAllRules(dto))); }
    @PostMapping("/rules") public ResponseEntity<ApiResponse<DrRule>> createRule(@RequestBody DrRule data) { return ResponseEntity.ok(ApiResponse.success(service.createRule(data), "创建成功")); }
    @PatchMapping("/rules/{id}") public ResponseEntity<ApiResponse<DrRule>> updateRule(@PathVariable Integer id, @RequestBody DrRule data) { return ResponseEntity.ok(ApiResponse.success(service.updateRule(id, data), "更新成功")); }
    @DeleteMapping("/rules/{id}") public ResponseEntity<ApiResponse<Void>> removeRule(@PathVariable Integer id) { service.removeRule(id); return ResponseEntity.ok(ApiResponse.success(null, "删除成功")); }

    @GetMapping("/groups") public ResponseEntity<ApiResponse<PaginatedResult<DrGroup>>> findAllGroups(PaginationDto dto) { return ResponseEntity.ok(ApiResponse.success(service.findAllGroups(dto))); }
    @PostMapping("/groups") public ResponseEntity<ApiResponse<DrGroup>> createGroup(@RequestBody DrGroup data) { return ResponseEntity.ok(ApiResponse.success(service.createGroup(data), "创建成功")); }
    @PatchMapping("/groups/{id}") public ResponseEntity<ApiResponse<DrGroup>> updateGroup(@PathVariable Integer id, @RequestBody DrGroup data) { return ResponseEntity.ok(ApiResponse.success(service.updateGroup(id, data), "更新成功")); }
    @DeleteMapping("/groups/{id}") public ResponseEntity<ApiResponse<Void>> removeGroup(@PathVariable Integer id) { service.removeGroup(id); return ResponseEntity.ok(ApiResponse.success(null, "删除成功")); }

    @GetMapping("/carriers") public ResponseEntity<ApiResponse<PaginatedResult<DrCarrier>>> findAllCarriers(PaginationDto dto) { return ResponseEntity.ok(ApiResponse.success(service.findAllCarriers(dto))); }
    @PostMapping("/carriers") public ResponseEntity<ApiResponse<DrCarrier>> createCarrier(@RequestBody DrCarrier data) { return ResponseEntity.ok(ApiResponse.success(service.createCarrier(data), "创建成功")); }
    @PatchMapping("/carriers/{id}") public ResponseEntity<ApiResponse<DrCarrier>> updateCarrier(@PathVariable Integer id, @RequestBody DrCarrier data) { return ResponseEntity.ok(ApiResponse.success(service.updateCarrier(id, data), "更新成功")); }
    @DeleteMapping("/carriers/{id}") public ResponseEntity<ApiResponse<Void>> removeCarrier(@PathVariable Integer id) { service.removeCarrier(id); return ResponseEntity.ok(ApiResponse.success(null, "删除成功")); }

    @PostMapping("/reload") public ResponseEntity<ApiResponse<Void>> reload() { service.reload(); return ResponseEntity.ok(ApiResponse.success(null, "重载成功")); }
    @GetMapping("/stats") public ResponseEntity<ApiResponse<Map<String, Object>>> getStats() { return ResponseEntity.ok(ApiResponse.success(service.getStats())); }
    @GetMapping("/gw-status") public ResponseEntity<ApiResponse<Object>> getGwStatus() { return ResponseEntity.ok(ApiResponse.success(service.getGwStatus())); }
    @GetMapping("/carrier-status") public ResponseEntity<ApiResponse<Object>> getCarrierStatus() { return ResponseEntity.ok(ApiResponse.success(service.getCarrierStatus())); }
}
